platform/ios/src/MGLScaleBarView.m

```diff
@@ -0,0 +1,387 @@
+#import <Mapbox/Mapbox.h>
+#import "MGLScaleBarView.h"
+
+
+@class MGLScaleBarLabel;
+
+@interface MGLScaleBarView()
+@property (nonatomic, assign, getter=isVisible) BOOL visible;
+@property (nonatomic) NSArray<MGLScaleBarLabel *> *labels;
+@property (nonatomic) NSArray<UIView *> *bars;
+@property (nonatomic) UIView *containerView;
+@property (nonatomic) MGLDistanceFormatter *formatter;
+@property (nonatomic) NSArray<NSArray<NSNumber *> *> *metricTable;
+@property (nonatomic) NSArray<NSArray<NSNumber *> *> *imperialTable;
+@property (nonatomic) NSArray<NSNumber*> *row;
+@property (nonatomic) UIColor *primaryColor;
+@property (nonatomic) UIColor *secondaryColor;
+@property (nonatomic) CALayer *borderLayer;
+@property (nonatomic, assign) CGFloat borderWidth;
+@end
+
+static const CGFloat MGLBarHeight = 4;
+static const CGFloat MGLFeetPerMeter = 3.28084;
+static const CGFloat MGLFeetPerMile = 5280;
+
+@interface MGLScaleBarLabel : UILabel
+@end
+
+@implementation MGLScaleBarLabel
+
+- (void)drawTextInRect:(CGRect)rect {
+    CGSize shadowOffset = self.shadowOffset;
+    UIColor *textColor = self.textColor;
+    
+    CGContextRef context = UIGraphicsGetCurrentContext();
+    CGContextSetLineWidth(context, 1);
+    CGContextSetLineJoin(context, kCGLineJoinRound);
+    
+    CGContextSetTextDrawingMode(context, kCGTextStroke);
+    self.textColor = [UIColor whiteColor];
+    [super drawTextInRect:rect];
+    
+    CGContextSetTextDrawingMode(context, kCGTextFill);
+    self.textColor = textColor;
+    self.shadowOffset = CGSizeMake(0, 0);
+    [super drawTextInRect:rect];
+    
+    self.shadowOffset = shadowOffset;
+}
+
+@end
+
+@implementation MGLScaleBarView
+
+- (instancetype)initWithCoder:(NSCoder *)decoder {
+    if (self = [super initWithCoder:decoder]) {
+        [self commonInit];
+    }
+    return self;
+}
+
+- (instancetype)initWithFrame:(CGRect)frame {
+    if (self = [super initWithFrame:frame]) {
+        [self commonInit];
+    }
+    return self;
+}
+
+- (void)commonInit {
+    _visible = NO;
+    _primaryColor = [UIColor colorWithRed:18.0/255.0 green:45.0/255.0 blue:17.0/255.0 alpha:1];
+    _secondaryColor = [UIColor colorWithRed:247.0/255.0 green:247.0/255.0 blue:247.0/255.0 alpha:1];
+    _borderWidth = 1.0f;
+    
+    self.clipsToBounds = NO;
+    self.alpha = 0;
+    
+    _containerView = [[UIView alloc] init];
+    _containerView.clipsToBounds = YES;
+    _containerView.backgroundColor = self.secondaryColor;
+    [self addSubview:_containerView];
+    
+    _borderLayer = [CAShapeLayer layer];
+    _borderLayer.borderColor = [self.primaryColor CGColor];
+    _borderLayer.borderWidth = 1.0f / [[UIScreen mainScreen] scale];
+    
+    [_containerView.layer addSublayer:_borderLayer];
+    
+    _formatter = [[MGLDistanceFormatter alloc] init];
+    _row = @[@0, @0];
+    
+    // @[meters, numberOfBars]
+    _metricTable = @[@[@1, @2],
```

Comment:
There are two important points here:

* The tables never change, so they should be allocated statically. A static C structure would be best; if an NSArray is really necessary, it should be created inside a dispatch_once block.
* The table is already sorted by distance, so the valid rows are always contiguous. Therefore, we can represent the valid rows as a range of indices into the static table rather than allocating a new subarray each time we relayout.

This code doesn't benefit from NSArray's mutability or its ability to hold objects.